feat: export tagged library as csv (TEMP)

rpenido opened the pull request.

rpenido committed:
feat: export tagged library as csv

rpenido committed:
Merge branch 'rpenido/fal-3610-download-course-tag-spreadsheet' into rpenido/fal-3611-download-library-tag-spreadsheet

rpenido committed:
Merge branch 'rpenido/fal-3610-download-course-tag-spreadsheet' into rpenido/fal-3611-download-library-tag-spreadsheet

rpenido committed:
Merge branch 'rpenido/fal-3610-download-course-tag-spreadsheet' into rpenido/fal-3611-download-library-tag-spreadsheet

rpenido committed:
fix: rebasing

rpenido committed:
fix: rename variable to fix typing

rpenido committed:
fix: pylint

rpenido committed:
Merge branch 'rpenido/fal-3610-download-course-tag-spreadsheet' into rpenido/fal-3611-download-library-tag-spreadsheet

rpenido committed:
refactor: changing condition checking

rpenido committed:
fix: check invalidkey

github-actions[bot] committed:
feat: Upgrade Python dependency ora2 (#34244)

Commit generated by workflow `openedx/edx-platform/.github/workflows/upgrade-one-python-dependency.yml@refs/heads/master`

Co-authored-by: jansenk <jansenk@users.noreply.github.com>
Co-authored-by: Jansen Kantor <jkantor@edx.org>

pomegranited commented on openedx/core/djangoapps/content_tagging/rest_api/v1/objecttag_export_helpers.py:
```diff
@@ -38,51 +42,136 @@ def iterate_with_level(
-    content_key: LearningContextKey,
-    object_tag_cache: ObjectTagByObjectIdDict,
-) -> TaggedContent:
+def get_course_tagged_object_and_children(
```
Could you prefix this function (and the other internal functions in the file) with an underscore, so we know that they're not meant for use outside of this helper file?

rpenido replied:
Done: a153ec92eba0e2250e58a15168873bcfc120dcd7

pomegranited commented on openedx/core/djangoapps/content_tagging/rest_api/v1/objecttag_export_helpers.py:
```diff
@@ -38,51 +42,136 @@ def iterate_with_level(
-    display_name = course.display_name_with_default
-    course_id = str(course.id)
+    course = store.get_course(course_key)
+    assert course is not None
```
Need to catch `AssertionFailed` somewhere -- maybe in the view?

rpenido replied:
I changed it to raise a `ValueError` and handle it in the view here: ed547bea0ff8c023dd68a18b17ca6b6b0f77da04

pomegranited commented on openedx/core/djangoapps/content_tagging/rest_api/v1/objecttag_export_helpers.py:
```diff
@@ -38,51 +42,136 @@ def iterate_with_level(
+    Returns a TaggedContent with library metadata with its tags, and its children.
+    """
+    library = library_api.get_library(library_key)
+    assert library is not None
```
Same -- `AssertionError` should be caught in the view.

rpenido replied:
Fixed: ed547bea0ff8c023dd68a18b17ca6b6b0f77da04

pomegranited commented on openedx/core/djangoapps/content_tagging/rest_api/v1/objecttag_export_helpers.py:
```diff
@@ -38,51 +42,136 @@ def iterate_with_level(
+    elif isinstance(content_key, LibraryLocatorV2):
+        tagged_content, children = get_library_tagged_object_and_children(
+            content_key, object_tag_cache
+        )
```
Up here we can decide which method to use when fetching child blocks, so we don't have to do the type checks in the loop below:

```suggestion
    if isinstance(content_key, CourseKey):
        tagged_content, children = get_course_tagged_object_and_children(
            content_key, object_tag_cache
        )
        get_tagged_children = get_xblock_tagged_object_and_children
    elif isinstance(content_key, LibraryLocatorV2):
        tagged_content, children = get_library_tagged_object_and_children(
            content_key, object_tag_cache
        )
        get_tagged_children = get_library_block_tagged_object
```

But you'll need to remove the `store` parameter from `get_xblock_tagged_object_and_children` and just call this again (it's ok to do this, the modulestore is globally cached):

```
    store = modulestore()
```

pomegranited commented on openedx/core/djangoapps/content_tagging/rest_api/v1/objecttag_export_helpers.py:
```diff
@@ -38,51 +42,136 @@ def iterate_with_level(
                 )
-                tagged_block.children.append(tagged_child)
+            else:
+                raise NotImplementedError(f"Invalid child: {type(child)} -> {child}")
```
Now we can use the `get_tagged_children` variable set in my suggestion above:

```suggestion
            tagged_child, child_children = get_tagged_children(child, object_tag_cache)
```

rpenido replied:
Done: 610c01aa3c6e196e1c987d7106cf9db2655f6ff9

At first, I thought we would still need a type guard before calling the function, ~~but the types were inferred just right~~.

The code is way cleaner now! Thank you!

Edit: My lsp didn't found any issues but our static type checker didn't like it. It uses the first variable assignment to define the type:
```
openedx/core/djangoapps/content_tagging/rest_api/v1/objecttag_export_helpers.py:150: error: Incompatible types in assignment (expression has type "Callable[[LibraryXBlockMetadata, Dict[str, Dict[int, List[Any]]]], Tuple[TaggedContent, None]]", variable has type "Callable[[UsageKey, Dict[str, Dict[int, List[Any]]]], Tuple[TaggedContent, List[Any]]]")  [assignment]
```

I did an explicit typing here: 40124816b4d33bb820e6269f544885a3a51af645
Not pretty, but I think it is better than disable the type checking. Do you know a better way to do this?

pomegranited replied:
@rpenido sorry, I don't know a better way to do this.. but I agree, having an ugly type definition is better than disabling type checking.

pomegranited commented on openedx/core/djangoapps/content_tagging/rest_api/v1/views.py:
```diff
@@ -141,7 +141,6 @@ def orgs(self, request, **_kwargs) -> Response:
 class ObjectTagOrgView(ObjectTagView):
     """
     View to create and retrieve ObjectTags for a provided Object ID (object_id).
-    This view extends the ObjectTagView to add Organization filters for the results.
```
Why remove this comment? I think it's still relevant.

rpenido replied:
Sorry! This was unintended and slipped through my check. Fixed: df06d9a891e77cb24a97df24661f7798d0f2bfa7

pomegranited reviewed (changes requested):
Hi @rpenido ! This is working well, but I've suggested a few changes here.

Thank you for splitting this code into an internal PR, it really helped me see what was changed for libraries :)

michaelroytman committed:
temp: add supplemental logging to debug IDV issues (#34248)

This commit adds some supplemental, more verbose logging to the results_callback view in the verify_student Djangoapp. This endpoint is called by identity verification (IDV) providers to POST an IDV review to edX.

We are experiencing issues with receiving IDV reviews from our IDV provider, and these logs will help us diagnose whether there is an issue in edX's systems.

These logs will be removed after our investigation is complete.

rpenido committed:
fix: better error handling

rpenido committed:
docs: reverting unintended change

rpenido committed:
refactor: change functions to private

rpenido committed:
refactor: cleaning build_object_tree_with_objecttags function

rpenido committed:
fix: explicit typing

rpenido committed:
feat: export all course tags as csv (#34091)

christopappas committed:
feat: make FA form error messaging more descript (#34247)

* feat: make FA form error messaging more descript

* chore: quality

* fix: split up tests

* fix: make test more specific

* chore: fix comment typo

* chore: fix comment typo

rpenido committed:
Merge branch 'master' into rpenido/fal-3611-download-library-tag-spreadsheet

rpenido committed:
fix: cleaning some types

rpenido committed:
docs: fix comment

pomegranited committed:
perf: reduce number of queries for content tagging endpoints (#34200)

rpenido committed:
Merge branch 'master' into rpenido/fal-3611-download-library-tag-spreadsheet

rpenido committed:
fix: typo

rpenido commented:
Closed in favor of https://github.com/openedx/edx-platform/pull/34246
